src/transport-manager.js

```diff
@@ -150,6 +158,9 @@ class TransportManager {
         const listener = transport.createListener({}, this.onConnection)
         this._listeners.get(key).push(listener)
 
+        // Track listen events
+        listener.on('listening', () => this.emit('listening'))
```

Comment:
I don't think the TransportManager needs to emit events. 

If we watch the `listening` and `close` events for each transport listener, instead of bubbling those events we can have the TransportManager itself create and seal the updated Peer Record. The TransportManager should always know when a listeners status changes, so pushing updates from here makes sense.

1. A transport begins/stops listening on an address
1. TransportManager sees the change and generates a new peer record and stores it the address book
1. Storing the new record triggers the multiaddr changes
1. Identify sees the multiaddr change event from the peer store for self and sends out a push

If we then store our protocols in the peer store instead of the upgrader, the identify service can also subscribe to those which keeps the api consistent.

Reply:
> If we then store our protocols in the peer store instead of the upgrader, the identify service can also subscribe to those which keeps the api consistent.

That I totally agree, and I will follow up a PR to store the protocols in protocol and subscribe to changes

Reply:
The advantage to internalizing the address change to the transport manager is that it doesn't affect the API on any system at this point so making that change is low risk. I think the Address Manager owning this makes sense, but I'd like to see a design before we change those APIs, to avoid needing to change them in the future.

The Address Manager should also own our observed addresses, tracking them and scoring them accordingly. How should observed addresses flow into the Address Manager? Does it affect the API?

Reply:
Ok, so I went with your suggestion for now. I was not expecting to break the API adding this to the AddressManager as it would be basically the same as now in the TransportManager  + TransportManager being an event emitter.

Regarding the observed addresses I agree. So, I will start for creating an issue and we can discuss this when it is relevant